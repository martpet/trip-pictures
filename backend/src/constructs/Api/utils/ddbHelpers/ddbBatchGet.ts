import {
  BatchGetCommand,
  DDB_BATCHGET_MAX_SIZE,
  DynamoDBClient,
  DynamoDBDocument,
  TableOptions,
} from 'lambda-layer';

const ddbDocClient = DynamoDBDocument.from(new DynamoDBClient({}));

type Props = {
  tableOptions: TableOptions;
  pkValues: string[];
};

export const ddbBatchGet = async <TableItem>({ tableOptions, pkValues }: Props) => {
  const { tableName, partitionKey } = tableOptions;
  const requests = [];
  let result: TableItem[] = [];

  for (let i = 0; i < pkValues.length; i += DDB_BATCHGET_MAX_SIZE) {
    const chunk = pkValues.slice(i, i + DDB_BATCHGET_MAX_SIZE);
    const keys = chunk.map((pkValue) => ({ [partitionKey.name]: pkValue }));
    requests.push(get(keys));
  }

  async function get(Keys: Record<string, unknown>[]) {
    const input = { RequestItems: { [tableName]: { Keys } } };
    const command = new BatchGetCommand(input);
    const { Responses, UnprocessedKeys } = await ddbDocClient.send(command);
    const items = Responses?.[tableName] as TableItem[] | undefined;
    const nextKeys = UnprocessedKeys?.[tableName]?.Keys;

    if (items) result = result.concat(items);
    if (nextKeys) await get(nextKeys);
  }

  await Promise.all(requests);

  return result;
};
