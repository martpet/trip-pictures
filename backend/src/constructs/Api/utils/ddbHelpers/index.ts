export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
