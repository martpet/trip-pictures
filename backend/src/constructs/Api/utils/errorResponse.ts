import { APIGatewayProxyResult } from 'aws-lambda';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

import { ApiErrorResponseBody } from '~/types';

type BaseOptions = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<BaseOptions, 'error'> & {
  exposeError: true;
};

type WithoutExposedError = BaseOptions & {
  exposeError?: never | false;
};

type ErrorResponseOptions = WithExposedError | WithoutExposedError;

export const errorResponse = (
  traceId: string,
  options: ErrorResponseOptions = {}
): APIGatewayProxyResult => {
  const {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    exposeError,
  } = options;

  const { envName } = globalLambdaProps;

  const errObj: ApiErrorResponseBody['error'] = {
    statusCode,
    message: getReasonPhrase(statusCode),
    traceId,
  };

  if (description) {
    errObj.description = description;
  }

  if (error && (exposeError ?? envName !== 'production')) {
    errObj.error = String(error);
  }

  const body: ApiErrorResponseBody = {
    error: errObj,
  };

  console.error({ traceId, error });

  return {
    statusCode,
    body: JSON.stringify(body),
  };
};
