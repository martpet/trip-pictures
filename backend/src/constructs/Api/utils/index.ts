export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
