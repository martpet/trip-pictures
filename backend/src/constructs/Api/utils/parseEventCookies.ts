import { APIGatewayProxyEventV2 } from 'aws-lambda';
import cookie from 'cookie';

import { Cookies, PathWithCookies } from '../types';

export const parseEventCookies = <T extends PathWithCookies>(
  event: APIGatewayProxyEventV2
): Cookies<T> => {
  try {
    const cookieHeaderString = event.cookies?.join(';');
    if (!cookieHeaderString) throw new Error();
    return cookie.parse(cookieHeaderString);
  } catch (e) {
    return {};
  }
};
