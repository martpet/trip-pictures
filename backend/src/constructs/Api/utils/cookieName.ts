import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
