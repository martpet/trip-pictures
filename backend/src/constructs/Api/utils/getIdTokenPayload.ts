import { IdTokenPayload } from '~/constructs/Api/types';

export const getIdTokenPayload = async (idToken: string): Promise<IdTokenPayload> => {
  const camelcaseKeys = (await import('camelcase-keys')).default;
  const payload = JSON.parse(Buffer.from(idToken.split('.')[1], 'base64').toString());

  return camelcaseKeys(payload);
};
