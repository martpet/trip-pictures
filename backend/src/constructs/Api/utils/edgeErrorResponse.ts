import { getReasonPhrase } from 'http-status-codes';

import { LambdaEdgeResponse } from '~/constructs/Api/types';
import { errorResponse } from '~/constructs/Api/utils/errorResponse';

export const edgeErrorResponse = (
  ...errorResponseArgs: Parameters<typeof errorResponse>
): LambdaEdgeResponse => {
  const { statusCode, body } = errorResponse(...errorResponseArgs);

  const response: LambdaEdgeResponse = {
    status: statusCode,
    statusDescription: getReasonPhrase(statusCode),
    headers: { 'content-type': [{ value: 'application/json' }] },
    body,
  };

  return response;
};
