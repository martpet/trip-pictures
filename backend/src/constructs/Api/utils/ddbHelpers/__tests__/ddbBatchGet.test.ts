import { BatchGetCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import { itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';
import { DDB_BATCHGET_MAX_SIZE } from '~/consts';

import { ddbBatchGet } from '../ddbBatchGet';

const ddbMock = mockClient(DynamoDBDocumentClient);

const tableOptions = {
  tableName: 'dummyTableName',
  partitionKey: { name: 'dummyPartitionKeyName' },
};

const args = [
  {
    tableOptions,
    pkValues: ['dummyPkValue'],
  },
] as Parameters<typeof ddbBatchGet>;

beforeEach(() => {
  ddbMock.reset();

  ddbMock.on(BatchGetCommand).resolves({
    Responses: {
      [tableOptions.tableName]: [{ dummyResponseKey: 'dummyResponseValue' }],
    },
  });
});

describe('ddbBatchGet', () => {
  itSendsAwsCommand(BatchGetCommand, ddbMock, ddbBatchGet, args);
  itResolves(ddbBatchGet, args);

  describe(`when called with ${DDB_BATCHGET_MAX_SIZE + 1} items`, () => {
    const argsClone = structuredClone(args);
    argsClone[0].pkValues = Array(DDB_BATCHGET_MAX_SIZE + 1);

    it('sends "BatchGetCommand" 2 times', async () => {
      await ddbBatchGet(...argsClone);
      expect(ddbMock.commandCalls(BatchGetCommand).length).toBe(2);
    });
  });

  describe('when "BatchGetCommand" response is missing "Responses"', () => {
    beforeEach(() => {
      ddbMock.on(BatchGetCommand).resolves({});
    });
    itResolves(ddbBatchGet, args);
  });

  describe('when "BatchGetCommand" response has "UnprocessedKeys"', () => {
    beforeEach(() => {
      ddbMock
        .on(BatchGetCommand)
        .resolvesOnce({
          Responses: {
            [tableOptions.tableName]: [{ dummyResponseKey: 'dummyResponseValue1' }],
          },
          UnprocessedKeys: {
            [tableOptions.tableName]: {
              Keys: [{ dummyUnprocessedKeyProp: 'dummyUnprocessedKeyValue' }],
            },
          },
        })
        .resolvesOnce({
          Responses: {
            [tableOptions.tableName]: [{ dummyResponseKey: 'dummyResponseValue2' }],
          },
        });
    });

    it('sends "BatchGetCommand" a second time with correct args', async () => {
      await ddbBatchGet(...args);
      expect(ddbMock.commandCalls(BatchGetCommand)[1].args[0].input).toMatchSnapshot();
    });

    itResolves(ddbBatchGet, args);
  });
});
