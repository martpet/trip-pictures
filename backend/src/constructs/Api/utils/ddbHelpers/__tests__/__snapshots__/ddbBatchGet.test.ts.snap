// Vitest Snapshot v1

exports[`ddbBatchGet > resolves with a correct value 1`] = `
[
  {
    "dummyResponseKey": "dummyResponseValue",
  },
]
`;

exports[`ddbBatchGet > sends "BatchGetCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "RequestItems": {
    "dummyTableName": {
      "Keys": [
        {
          "dummyPartitionKeyName": "dummyPkValue",
        },
      ],
    },
  },
}
`;

exports[`ddbBatchGet > when "BatchGetCommand" response has "UnprocessedKeys" > resolves with a correct value 1`] = `
[
  {
    "dummyResponseKey": "dummyResponseValue1",
  },
  {
    "dummyResponseKey": "dummyResponseValue2",
  },
]
`;

exports[`ddbBatchGet > when "BatchGetCommand" response has "UnprocessedKeys" > sends "BatchGetCommand" a second time with correct args 1`] = `
{
  "RequestItems": {
    "dummyTableName": {
      "Keys": [
        {
          "dummyUnprocessedKeyProp": "dummyUnprocessedKeyValue",
        },
      ],
    },
  },
}
`;

exports[`ddbBatchGet > when "BatchGetCommand" response is missing "Responses" > resolves with a correct value 1`] = `[]`;
